### ISSUE TITLE
expect {}.to change obj, :msg not behaving as expected

### ISSUE THREAD

**Issue by knovoselic** [rspec/rspec-expectations#791]
Hi. I wrote a spec like this:

    expect {
      put :add_device, {device_name: 'test'}
    }.not_to change user, :devices

This test should fail because `add_device` is changing `user.devices` (by adding another object into the relation), but it's passing.
`User` is Mongoid model, and `devices` is embedded relation in `User` model.

After a little debugging with pry, I've found a object in `rspec/matchers/built_in/change.rb` on line 50  called `@change_details`. It contains two fields of interest - `actual_before` and `actual_after`. After comparing the two it seems that both point to the same object in memory (array), so `actual_after` got updated after new `Device` was added into User's embedded relation.
Here's the proof that they are pointing to the same object:

    [10] pry(#<RSpec::Matchers::BuiltIn::Change>)> @change_details.actual_before.object_id
    => 70163135333520
    [11] pry(#<RSpec::Matchers::BuiltIn::Change>)> @change_details.actual_after.object_id
    => 70163135333520

So here's my question: is this expected behaviour or a bug? If it's a bug, it should be easy to fix by calling `#dup` (or maybe `#clone`?) while retrieving `actual_before`. I can create pull request.
If I change spec like this, test works (fails) as expected:

    expect {
      put :add_device, {device_name: 'test'}
    }.not_to change user.devices, :dup

**Comment by myronmarston** [comment]
We already `dup` arrays and other enumerables to deal with this:

https://github.com/rspec/rspec-expectations/blob/54a57106788043284d9ebc7bba666a87fe234ba6/lib/rspec/matchers/built_in/change.rb#L329

So, my guess is that the mongoid relation isn't actually an array and in fact isn't `Enumerable` even though it may report itself to be an array when you inspect it.  Can you look into that?  If the mongoid relation is meant to be treated as an enumerable, it should include `Enumerable` so that libraries like RSpec know it can be treated that way.

**Comment by knovoselic** [comment]
Thanks for such a quick reply. I've debugged a little more and it seems that the problems is in === operator:

    [3] pry(#<RSpec::Matchers::BuiltIn::ChangeDetails>):1> val.is_a? Enumerable
    => true
    [4] pry(#<RSpec::Matchers::BuiltIn::ChangeDetails>):1> val.class
    => Array
    [5] pry(#<RSpec::Matchers::BuiltIn::ChangeDetails>):1> val === Array
    => false
    [6] pry(#<RSpec::Matchers::BuiltIn::ChangeDetails>):1> val === Enumerable
    => false
    [7] pry(#<RSpec::Matchers::BuiltIn::ChangeDetails>):1> val === Device
    => false

Can you help me locate which definition of `===` operator is in use? I'm not able to see it using pry's step command. And `show-method`/`show-source` isn't working:

    [3] pry(#<RSpec::ExampleGroups::ApiV1UsersController::PUTDeviceStatistics::WithValidDeviceStatisticsParameter>)> show-method user.devices.===
    Error: Couldn't locate a definition for user.devices.===!
    [4] pry(#<RSpec::ExampleGroups::ApiV1UsersController::PUTDeviceStatistics::WithValidDeviceStatisticsParameter>)> show-source user.devices.===
    Error: Couldn't locate a definition for user.devices.===!

**Comment by myronmarston** [comment]
> Thanks for such a quick reply. I've debugged a little more and it seems that the problems is in === operator:

The `===` is working correctly there.  `===` is not a symmetric operator.  `Array === []` returns true but `[] === Array` returns false and that's how it is designed to work.  `===` defines a _category_ to which objects may or may not belong and is such is meant to be applied one way.

What makes you think that `===` is the problem here?

**Comment by knovoselic** [comment]
Oh, I'm still learning ruby/rails so I obviously had wrong idea about `===` operator. I didn't know that it's not a symmetric operator.

So if `val.is_a? Enumerable` is `true`, and `val.is_a? Array` is `true`, why isn't case working?
    [14] pry(#<RSpec::Matchers::BuiltIn::ChangeDetails>)> val.is_a? Enumerable
    => true
    [15] pry(#<RSpec::Matchers::BuiltIn::ChangeDetails>)> val.is_a? Array
    => true

Shouldn't the following return true?

    [16] pry(#<RSpec::Matchers::BuiltIn::ChangeDetails>)> Array === val
    => false
    [17] pry(#<RSpec::Matchers::BuiltIn::ChangeDetails>)> Enumerable === val
    => false

**Comment by myronmarston** [comment]
> Shouldn't the following return true?

Ideally, it would, but in some circumstances it might not.  Any object can define `is_a?` to return whatever value it wants (including lying about it being a type that it's not) where as `Array === val` calls a method on `Array` so you'd have to monkey patch `Array` to make that return `true` when given an object that's not really an array.  (That's actually why we tend to use `===` throughout RSpec: we are given all manner of user objects that may have redefined `is_a?`--such as `user.is_a?(:admin)`--and as such we can't trust it like we can trust `Class#===`).

My guess is that `val` is a proxy object that wraps an array and proxies method calls to it, including `is_a?`, so that it lies and claims to be an array when it's not really.  You can check by doing `val.method(:is_a?).source_location` and `val.method(:is_a?).owner` to see where `is_a?` is defined and in what module it is defined.

**Comment by knovoselic** [comment]
Thanks again, this is very useful to me, I'm learning a lot.

Does the following mean that `is_a?` and `===` aren't overridden?

    [6] pry(#<RSpec::Matchers::BuiltIn::ChangeDetails>)> val.method(:is_a?).source_location
    => nil
    [7] pry(#<RSpec::Matchers::BuiltIn::ChangeDetails>)> val.method(:is_a?).owner
    => Kernel
    [11] pry(#<RSpec::Matchers::BuiltIn::ChangeDetails>)> val.method(:===).owner
    => Kernel
    [12] pry(#<RSpec::Matchers::BuiltIn::ChangeDetails>)> val.method(:===).source_location
    => nil

**Comment by myronmarston** [comment]
> Does the following mean that is_a? and === aren't overridden?

Yes, I think so.

Can you put together a reproducible example we can play with?

**Comment by knovoselic** [comment]
Will do first thing tomorrow. I'm going off for now. Thanks for helping me with this.

**Comment by knovoselic** [comment]
@myronmarston here's a reproducible example: https://github.com/knovoselic/rspec-expectations-change-bug-repro

Just run User specs - all three should pass, but the first one is failing.

**Comment by myronmarston** [comment]
Thanks for the clear repro case, @knovoselic!

I confirmed that the problem is indeed that the relation isn't an `Enumerable` (although it does proxy to one).  I found that it can be fixed if you add this line:

``` ruby
Mongoid::Relations::Many.include Enumerable
```

Of course, monkey patching like this isn't ideal, and it seems to me that mongoid itself should make the `Many` proxy enumerable.  Besides the fact that it will enable RSpec to realize it's a collection and treat it as one, it will also add many useful enumeration methods and may help it integrate with other libraries that similarly use `Enumerable` as a way to detect "can be treated like a collection".

Can you open an issue (or PR) with Mongoid to have `Mongoid::Relations::Many` include `Enumerable`?

**Comment by cupakromer** [comment]
To add a 2nd take on the issue. It seems the relation is a proxy (`Mongoid::Relations::Embedded::Many` < `Mongoid::Relations::Many` < `Mongoid::Relations::Proxy`) which delegates to an `Array` object:

```ruby
user.devices.class
# => Array
user.devices.is_a?(Array)
# => true
user.devices.kind_of?(Array)
# => true
Array === user.devices
# => false
user.devices === Array
# => false
user.devices < Array
# => nil
```

If you look at the implementation of `Module.===` you will see:

```c
static VALUE
rb_mod_eqq(VALUE mod, VALUE arg)
{
    return rb_obj_is_kind_of(arg, mod);
}
```

Unfortunately, while `rb_obj_is_kind_of` is the C-function name of `Object.kind_of?` it seems that this lower level function is not proxied through. Thus, it is un-able to proxy the sub-class check performed by `===` as has already been discussed.

In other words, Mongoid has put a bit of effort into making it appear that the relation is an `Array`, but this falls short in some critical checks. As @myronmarston stated, this is something that needs to be fixed on the Mongoid side.

**Comment by myronmarston** [comment]
> Unfortunately, while `rb_obj_is_kind_of` is the C-function name of `Object.kind_of?` it seems that this lower level function is not proxied through.

The reason it's not proxied through has nothing to do with it being a low-level C function...it's that the receiver in `Array === user.devices` is `Array`, not the `user.devices` proxy object that proxies method calls.

**Comment by JonRowe** [comment]
Is there anything we can do here @myronmarston? Looks like this is an issue with Mongoid here?

**Comment by myronmarston** [comment]
> Is there anything we can do here @myronmarston? Looks like this is an issue with Mongoid here?

Yep.  Closing.

**Comment by knovoselic** [comment]
Thanks for looking into this. I've opened an issue with Mongoid: https://github.com/mongoid/mongoid/issues/4077
